File: system-test/java/src/main/java/com/mycompany/myshop/testkit/driver/port/dtos/error/SystemError.java
package com.mycompany.myshop.testkit.driver.port.dtos.error;

import lombok.Builder;
import lombok.Data;
import lombok.Getter;

import java.util.Arrays;
import java.util.List;

@Data
@Builder
public class SystemError {
    private final String message;
    private final List<FieldError> fields;

    public static SystemError of(String message) {
        return SystemError.builder()
                .message(message)
                .build();
    }

    public static SystemError of(String message, FieldError... fieldErrors) {
        return SystemError.builder()
                .message(message)
                .fields(Arrays.asList(fieldErrors))
                .build();
    }

    public static SystemError of(String message, List<FieldError> fieldErrors) {
        return SystemError.builder()
                .message(message)
                .fields(fieldErrors)
                .build();
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("SystemError{message='").append(message).append("'");

        if (fields != null && !fields.isEmpty()) {
            sb.append(", fieldErrors=[");
            for (int i = 0; i < fields.size(); i++) {
                if (i > 0) sb.append(", ");
                sb.append(fields.get(i));
            }
            sb.append("]");
        }

        sb.append("}");
        return sb.toString();
    }

    @Getter
    public static class FieldError {
        private final String field;
        private final String message;
        private final String code;

        public FieldError(String field, String message, String code) {
            this.field = field;
            this.message = message;
            this.code = code;
        }

        public FieldError(String field, String message) {
            this(field, message, null);
        }

        @Override
        public String toString() {
            StringBuilder sb = new StringBuilder();
            sb.append("FieldError{field='").append(field).append("'");
            sb.append(", message='").append(message).append("'");
            if (code != null) {
                sb.append(", code='").append(code).append("'");
            }
            sb.append("}");
            return sb.toString();
        }
    }
}

File: system-test/dotnet/Driver.Port/Dtos/Error/SystemError.cs
namespace Driver.Port.Dtos.Error;

public class SystemError
{
    public string Message { get; private set; }
    public IReadOnlyList<FieldError>? Fields { get; private set; }

    private SystemError(string message, IReadOnlyList<FieldError>? fields = null)
    {
        Message = message;
        Fields = fields;
    }

    public static SystemError Of(string message)
    {
        return new SystemError(message);
    }

    public static SystemError Of(string message, params FieldError[] fieldErrors)
    {
        return new SystemError(message, fieldErrors.ToList().AsReadOnly());
    }

    public static SystemError Of(string message, IReadOnlyList<FieldError> fieldErrors)
    {
        return new SystemError(message, fieldErrors);
    }

    public override string ToString()
    {
        var sb = new System.Text.StringBuilder();
        sb.Append($"SystemError{{message='{Message}'");

        if (Fields != null && Fields.Any())
        {
            sb.Append(", fieldErrors=[");
            for (int i = 0; i < Fields.Count; i++)
            {
                if (i > 0) sb.Append(", ");
                sb.Append(Fields[i]);
            }
            sb.Append("]");
        }

        sb.Append("}");
        return sb.ToString();
    }

    public class FieldError
    {
        public string Field { get; }
        public string Message { get; }
        public string? Code { get; }

        public FieldError(string field, string message, string? code = null)
        {
            Field = field;
            Message = message;
            Code = code;
        }

        public override string ToString()
        {
            var sb = new System.Text.StringBuilder();
            sb.Append($"FieldError{{field='{Field}'");
            sb.Append($", message='{Message}'");
            if (!string.IsNullOrEmpty(Code))
            {
                sb.Append($", code='{Code}'");
            }
            sb.Append("}");
            return sb.ToString();
        }
    }
}

